lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp

```diff
@@ -140,6 +140,54 @@ class StableHLOToTTIRTransposeOpConversionPattern
         srcOp, outputTensor.getType(), Value(adaptor.getOperand()),
         Value(outputTensor), adaptor.getPermutation()[0],
         adaptor.getPermutation()[1],
+
+        rewriter.getArrayAttr(
+            SmallVector<Attribute>(adaptor.getOperands().size() + 1,
+                                   rewriter.getAttr<OperandConstraintAttr>(
+                                       OperandConstraint::AnyDeviceTile))));
+    return success();
+  }
+};
+
+class StableHLOToTTIRDotGeneralOpConversionPattern
+    : public OpConversionPattern<mlir::stablehlo::DotGeneralOp> {
+  using OpConversionPattern<mlir::stablehlo::DotGeneralOp>::OpConversionPattern;
+
+public:
+  LogicalResult
+  matchAndRewrite(mlir::stablehlo::DotGeneralOp srcOp,
+                  mlir::stablehlo::DotGeneralOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto outputType = mlir::cast<RankedTensorType>(srcOp.getResult().getType());
+    auto outputTensor = rewriter.create<tensor::EmptyOp>(
+        srcOp.getLoc(), outputType.getShape(), outputType.getElementType());
+
+    // This is a basic version that can only work for cases that can be directly
+    // converted to matmul. The op should be extended as other ops such as
+    // ttir.permute and ttir.broadcast_in_dim become available.
+
+    ::mlir::stablehlo::DotDimensionNumbersAttr dimensions =
+        adaptor.getDotDimensionNumbers();
+    if (dimensions.getLhsContractingDimensions().size()) {
+      assert(dimensions.getLhsContractingDimensions()[0] == 1 &&
+             "ttir dot_general only supports matmul operation.");
+    }
+
+    if (dimensions.getRhsContractingDimensions().size()) {
+      assert(dimensions.getRhsContractingDimensions()[0] == 0 &&
+             "ttir dot_general only supports matmul operation.");
+    }
+
+    assert(dimensions.getLhsBatchingDimensions().empty() &&
+           "ttir dot_general only supports matmul operation.");
+
+    assert(dimensions.getRhsBatchingDimensions().empty() &&
+           "ttir dot_general only supports matmul operation.");
```

Comment:
I think at this stage of our compiler, if the input itself is correct, most internal errors are catastrophic i.e. the compiler cannot recover from it, that's why we use assertions. At the same time, we should provide user with a meaningful feedback about why the error occurred, hence an error message is very important. The issue with using `failure()` alone as it is now in `checkBasicLegality()` is that it doesn't inform about the cause of the failure. 

I suppose one can use `failure()` along with a debug message e.g. `LLVM_DEBUG()`, but then the user will have to open the debug log and find the cause of the error (which is fine for some cases). Hence, it is very common to use `llvm_unreachable("error message") `and `LLVM_ASSERT(condition && "error message")` are the most common LLVM based compilers. 

If the error is due to incorrect input, in that case `failure()` is a suitable option, but should include a reason so user can correct the input and try again.

I think this article best describes these points: 
http://onnx.ai/onnx-mlir/ErrorHandling.html

So to summarize I think one of the following ways is acceptable: 

1) `llvm_reachable()` or `LLVM_ASSERT()` with a brief message.
2) `op->emitError() << "message";`
3) `failure()` which is proceeded with `LLVM_DEBUG()` or `LLVM_WARN()`

Reply:
I ended up using a combination of `op->emitError() << "message"; `and `failure()` as it produces an unrecoverable failure as well as prints a debug message to user.

Reply:
Hey Usman, I totally agree that error message is also important, my main point was that assertions are not a good way of reporting errors to the user, their main usage is to catch code errors for developers. Assertions -> things you expect to never happen, otherwise it’s a bug. Errors -> things you expect to happen and come from input to the program, which is the case here. Even the link you posted supports this, you were asserting during input graph validation phase.

I like your investigation about proper error handling in llvm and the solution that you ended up using, we already use `op->emitError()` in other passes, it is basically a wrapper around `return failure()` which also prints error message.